pkg/envoy/ads/response.go

```diff
@@ -16,71 +20,91 @@ const (
 
 	log.Trace().Msgf("[%s] Creating response for proxy with SerialNumber=%s on Pod with UID=%s", xdsShortName, proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
 
-	discoveryResponse, err := s.newAggregatedDiscoveryResponse(proxy, req, cfg)
-	if err != nil {
-		log.Error().Err(err).Msgf("[%s] Failed to create response for proxy with SerialNumber=%s on Pod with UID=%s", xdsShortName, proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-		return err
+	handler, ok := s.xdsHandlers[tURI]
+	if !ok {
+		log.Error().Msgf("Responder for TypeUrl %s is not implemented", tURI)
+		return nil, errUnknownTypeURL
 	}
 
-	if err := (*server).Send(discoveryResponse); err != nil {
-		log.Error().Err(err).Msgf("[%s] Error sending to proxy with SerialNumber=%s on Pod with UID=%s", xdsShortName, proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-		return err
+	if s.cfg.IsDebugServerEnabled() {
+		s.trackXDSLog(proxy.GetCertificateCommonName(), tURI)
+	}
+
+	log.Trace().Msgf("Invoking handler for type %s, proxy Serial %s", tURI, proxy.GetCertificateSerialNumber())
+	resources, err := handler(s.catalog, proxy, req, cfg, s.certManager)
+	if err != nil {
+		log.Error().Err(err).Msgf("Responder executing handler for %s", tURI)
+		return nil, err
 	}
 
 	success = true // read by deferred function
-	return nil
+	return resources, nil
 }
 
-func (s *Server) sendAllResponses(proxy *envoy.Proxy, server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer, cfg configurator.Configurator) {
-	log.Trace().Msgf("A change announcement triggered *DS update for proxy with SerialNumber=%s on Pod with UID=%s", proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-
-	// Tracks the success of this full update of all its XDS paths. If a single XDS response path fails for this full update,
-	// the full updated will be considered as failed for metric purposes (success = false)
+// sendResponse takes a set of TypeURIs which will be called to generate the xDS resources
+// for, and will have them sent to the proxy server.
+// If no DiscoveryRequest is passed, an empty one for the TypeURI is created
+func (s *Server) sendResponse(typeURIsToSend mapset.Set,
+	proxy *envoy.Proxy,
+	server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer,
+	request *xds_discovery.DiscoveryRequest,
+	cfg configurator.Configurator) error {
 	success := true
-	defer xdsPathTimeTrack(time.Now(), log.Info(), ADSUpdateStr, proxy.GetCertificateSerialNumber().String(), &success)
+	if typeURIsToSend.Cardinality() == len(envoy.XDSResponseOrder) {
+		defer xdsPathTimeTrack(time.Now(), log.Info(), ADSUpdateStr, proxy.GetCertificateSerialNumber().String(), &success)
+	}
+
+	// Temporary store for XDS resources
+	resourceMap := map[envoy.TypeURI][]types.Resource{}
 
-	// Order is important: CDS, EDS, LDS, RDS
+	// Order is important: CDS, EDS, LDS, RDS, SDS
```

Comment:
I'd exclude SDS from this list.
Also, the link to @kyessenov's `go-control-plane` comment below on xDS ordering is from 2018 - we should check to see whether 3 years later this is still the case.